### PR TITLE
Half float checkpoint

### PR DESCRIPTION
**Pull request by dcslin**
added half cpp backend, half tensor conversion, half tensor math operations, added half to linear, softmax layer, cnn layers, opt module. tested examples naive, module, train_cnn on mlp and cnn, refactored broadcast, added benchmark for fp16 vs fp32

### CODE REVIEW

**Comment by dcslin** [comment]
Usage example:
```
root@1c6aaef3db53:~/singa-hpmerge# PYTHONPATH=build/python/ python3 examples/cnn/train_cnn.py cnn mnist -m3 -pfloat16 -i 7
Starting Epoch 0:
Training loss = 0.061211, training accuracy = 0.752235
Evaluation accuracy = 0.944912, Elapsed Time = 6.867703s
Starting Epoch 1:
Training loss = 237.194122, training accuracy = 0.920124
Evaluation accuracy = 0.957131, Elapsed Time = 6.740274s
Starting Epoch 2:
Training loss = 167.028595, training accuracy = 0.944337
Evaluation accuracy = 0.970353, Elapsed Time = 6.833484s
root@1c6aaef3db53:~/singa-hpmerge# PYTHONPATH=build/python/ python3 examples/cnn/train_cnn.py cnn mnist -m3 -pfloat32 -i 7
Starting Epoch 0:
Training loss = 596.964600, training accuracy = 0.789421
Evaluation accuracy = 0.943209, Elapsed Time = 5.502895s
Starting Epoch 1:
Training loss = 234.664322, training accuracy = 0.920758
Evaluation accuracy = 0.960036, Elapsed Time = 5.506354s
Starting Epoch 2:
Training loss = 165.501694, training accuracy = 0.944454
Evaluation accuracy = 0.971254, Elapsed Time = 5.482749s
```

File: include/half.hpp [hunk 1]
```diff
@@ -0,0 +1,4575 @@
+// half - IEEE 754-based half-precision floating-point library.
+//
```
Comment (nudles) [inline include/half.hpp:2]: put it under singa/core/ ?
